In [1]:
from google.colab import drive

In [4]:
drive.mount('/content/drive')

Mounted at /content/drive


In [5]:
import pandas as pd

# export df to .csv
from google.colab import files

albums.csv

In [6]:
df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Data/spotify_albums.csv')

In [7]:
# Drop first column 
df = df.iloc[: , 1:]

In [8]:
# Show first 2 rows of the dataset:
df.head(2)

album_type               artist_id  \
0     single  3DiDSECUqqY1AuBP8qtaIa   
1      album  6s1pCNXcbdtQJlsnM1hRIA   

                                   available_markets  \
0  ['AD', 'AE', 'AR', 'AT', 'AU', 'BE', 'BG', 'BH...   
1  ['AD', 'AE', 'AR', 'AT', 'AU', 'BE', 'BG', 'BH...   

                                       external_urls  \
0  {'spotify': 'https://open.spotify.com/album/1g...   
1  {'spotify': 'https://open.spotify.com/album/4K...   

                                                href                      id  \
0  https://api.spotify.com/v1/albums/1gAM7M4rBwEb...  1gAM7M4rBwEbSPeAQR2nx1   
1  https://api.spotify.com/v1/albums/4KfJZV7WfolY...  4KfJZV7WfolYlxBzOTo66s   

                                              images  \
0  [{'height': 640, 'url': 'https://i.scdn.co/ima...   
1  [{'height': 640, 'url': 'https://i.scdn.co/ima...   

                                                name release_date  \
0                              If I Ain't Got You EP   2019-02-08   
1  Shostakovich Symphony No.5 - Four Romances on ...   2019-03-01   

  release_date_precision  total_tracks                track_id  \
0                    day             6  2iejTMy9XZ8Gaae0aQ2yl0   
1                    day             8  1WQfghEjszJJ4H8MAWrQ2C   

  track_name_prev                                   uri   type  
0        track_32  spotify:album:1gAM7M4rBwEbSPeAQR2nx1  album  
1        track_11  spotify:album:4KfJZV7WfolYlxBzOTo66s  album

In [ ]:
artist_id = df['artist_id']

for row in artist_id[0:50]:
    print(row)
    print(type(row))

In [ ]:
id = df['id']

for row in id[0:50]:
    print(row)
    print(type(row))

In [ ]:
track_id = df['track_id']

for row in track_id[0:50]:
    print(row)
    print(type(row))

Notable Columns:

In [ ]:
# 'images' column: contains image url
df['images'].iloc[0]

"[{'height': 640, 'url': 'https://i.scdn.co/image/5872e4d8fac4ef7552576d481b1d676189b4056a', 'width': 640}, {'height': 300, 'url': 'https://i.scdn.co/image/983855cfaf96ee784a176fff00ebbba541dcad04', 'width': 300}, {'height': 64, 'url': 'https://i.scdn.co/image/328e19dd8caaaaa827858875be19a82f47a9b96c', 'width': 64}]"

In [ ]:
# This column provides a url that opens up the Spotify website:
df['external_urls'].iloc[0]

"{'spotify': 'https://open.spotify.com/album/1gAM7M4rBwEbSPeAQR2nx1'}"

In [ ]:
# check null value for entire df:
df.isnull().values.any()

False

In [ ]:
# Stats for the whole table:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 75511 entries, 0 to 75510
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   album_type              75511 non-null  object
 1   artist_id               75511 non-null  object
 2   available_markets       75511 non-null  object
 3   external_urls           75511 non-null  object
 4   href                    75511 non-null  object
 5   id                      75511 non-null  object
 6   images                  75511 non-null  object
 7   name                    75511 non-null  object
 8   release_date            75511 non-null  object
 9   release_date_precision  75511 non-null  object
 10  total_tracks            75511 non-null  int64 
 11  track_id                75511 non-null  object
 12  track_name_prev         75511 non-null  object
 13  uri                     75511 non-null  object
 14  type                    75511 non-null  object
dtypes:

From above, we can see albums.csv has:
75,511 rows, 15 columns.

Stats for Individual columns:

In [ ]:
# Stats for column: 'total_tracks':
df['total_tracks'].describe()

count    75511.000000
mean         8.235807
std         11.669811
min          1.000000
25%          1.000000
50%          5.000000
75%         12.000000
max        977.000000
Name: total_tracks, dtype: float64

In [ ]:
# Out of curiosity, which row has the maximum 977 tracks??
res_df = df.loc[df['total_tracks'] == 977]
res_df

album_type  ...   type
18435  compilation  ...  album

[1 rows x 15 columns]

In [ ]:
# different 'album_type' values:
df['album_type'].unique()

array(['single', 'album', 'compilation'], dtype=object)

Export the cleaned file:


In [ ]:
# df.to_csv('clean_albums.csv') 
# files.download('clean_albums.csv')

artists.csv

In [12]:
df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Data/spotify_artists.csv')

In [13]:
# Drop first column 
df = df.iloc[: , 1:]

In [14]:
# check null value for entire df:
df.isnull().values.any()

True

Data Cleaning:

In [15]:
# Output rows with null values:
is_NaN = df.isnull()
row_has_NaN = is_NaN.any(axis=1)
rows_with_NaN = df[row_has_NaN]
rows_with_NaN

artist_popularity  followers genres                      id name  \
3836                 50       1142     []  4oPYazJJ1o4rWBrTw9lm40  NaN   

                    track_id track_name_prev    type  
3836  5OJZjoVoyNbzD24C7tkycW         track_7  artist

In [16]:
# Only one row has null value, looks like missing artist name. Decide to drop this row.
df = df.drop(3836)

In [17]:
# After dropping that one row, check null values for the entire df again:
df.isnull().values.any()

False

In [18]:
# Stats for the whole table:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 56128 entries, 0 to 56128
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   artist_popularity  56128 non-null  int64 
 1   followers          56128 non-null  int64 
 2   genres             56128 non-null  object
 3   id                 56128 non-null  object
 4   name               56128 non-null  object
 5   track_id           56128 non-null  object
 6   track_name_prev    56128 non-null  object
 7   type               56128 non-null  object
dtypes: int64(2), object(6)
memory usage: 3.9+ MB


After cleaning, artists.csv has: 56,128 rows, 8 columns.

In [19]:
# Show first 2 rows of the dataset:
df.head(2)

artist_popularity  followers  \
0                 44      23230   
1                 22        313   

                                              genres                      id  \
0  ['sertanejo', 'sertanejo pop', 'sertanejo trad...  4mGnpjhqgx4RUdsIJiURdo   
1                                                 []  1dLnVku4VQUOLswwDFvRc9   

             name                track_id track_name_prev    type  
0   Juliano Cezar  0wmDmAILuW9e2aRttkl4aC         track_9  artist  
1  The Grenadines  4wqwj0gA8qPZKLl5WVqXml        track_30  artist

In [ ]:
id = df['id']

for row in id[0:50]:
    print(row)
    print(type(row))

In [ ]:
track_id = df['track_id']

for row in track_id[0:50]:
    print(row)
    print(type(row))

In [ ]:
# Empty genres rows:
empty_genre_df = df.loc[df['genres'] == '[]']
empty_genre_df

artist_popularity  followers  ... track_name_prev    type
1                     22        313  ...        track_30  artist
5                     43         81  ...        track_15  artist
6                     34       8358  ...        track_43  artist
7                      7        158  ...        track_48  artist
8                     21         30  ...        track_73  artist
...                  ...        ...  ...             ...     ...
56107                 47          7  ...        track_45  artist
56110                 41        669  ...        track_26  artist
56117                 34        648  ...        track_72  artist
56118                 55        207  ...        track_76  artist
56122                 41       1282  ...         track_4  artist

[23539 rows x 8 columns]

Individual Column Exploration:

In [ ]:
df['artist_popularity'].describe()

count    56128.000000
mean        34.387169
std         16.917310
min          0.000000
25%         22.000000
50%         34.000000
75%         46.000000
max        100.000000
Name: artist_popularity, dtype: float64

In [ ]:
df['followers'].describe()

count    5.612800e+04
mean     7.796178e+04
std      5.942325e+05
min      0.000000e+00
25%      1.820000e+02
50%      1.734500e+03
75%      1.520325e+04
max      4.156169e+07
Name: followers, dtype: float64

In [ ]:
# different 'genres' values:
df['genres'].unique()

array(["['sertanejo', 'sertanejo pop', 'sertanejo tradicional', 'sertanejo universitario']",
       '[]', "['danish pop rock']", ...,
       "['dance pop', 'hip pop', 'pop', 'post-teen pop', 'r&b']",
       "['chilean rock', 'concepcion indie']",
       "['aggro chileno', 'andean', 'chilean rock', 'latin rock', 'nueva cancion']"],
      dtype=object)

In [ ]:
df['genres'].describe()

count     56128
unique    13314
top          []
freq      23539
Name: genres, dtype: object

Export the cleaned file:

In [ ]:
# df.to_csv('clean_artists.csv') 
# files.download('clean_artists.csv')

tracks.csv

In [42]:
df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Data/spotify_tracks.csv')

In [43]:
# Drop first column 
df = df.iloc[: , 1:]

In [44]:
# check null value for entire df:
df.isnull().values.any()

False

In [45]:
df.head()

acousticness                album_id  \
0         0.294  0D3QufeCudpQANOR7luqdr   
1         0.863  1bcqsH5UyTBzmh9YizdsBE   
2         0.750  4tKijjmxGClg4JOLAyo2qE   
3         0.763  6FeJF5r8roonnKraJxr4oB   
4         0.770  4tKijjmxGClg4JOLAyo2qE   

                                        analysis_url  \
0  https://api.spotify.com/v1/audio-analysis/5qlj...   
1  https://api.spotify.com/v1/audio-analysis/3VAX...   
2  https://api.spotify.com/v1/audio-analysis/1L3Y...   
3  https://api.spotify.com/v1/audio-analysis/6aCe...   
4  https://api.spotify.com/v1/audio-analysis/1Vo8...   

                   artists_id  \
0  ['3mxJuHRn2ZWD5OofvJtDZY']   
1  ['4xWMewm6CYMstu0sPgd9jJ']   
2  ['3hYaK5FF3YAglCj5HZgBnP']   
3  ['2KQsUB9DRBcJk17JWX1eXD']   
4  ['3hYaK5FF3YAglCj5HZgBnP']   

                                   available_markets country  danceability  \
0  ['AD', 'AE', 'AR', 'AT', 'AU', 'BE', 'BG', 'BH...      BE         0.698   
1  ['AD', 'AE', 'AR', 'AT', 'AU', 'BE', 'BG', 'BH...      BE         0.719   
2                                             ['GB']      BE         0.466   
3  ['AD', 'AE', 'AR', 'AT', 'AU', 'BE', 'BG', 'BH...      BE         0.719   
4                                             ['GB']      BE         0.460   

   disc_number  duration_ms  energy  ...  \
0          1.0     235584.0   0.606  ...   
1          1.0     656960.0   0.308  ...   
2          1.0     492840.0   0.931  ...   
3          1.0     316578.0   0.126  ...   
4          1.0     558880.0   0.942  ...   

                                         preview_url speechiness    tempo  \
0  https://p.scdn.co/mp3-preview/1b05a902da3a251d...      0.0262  115.018   
1  https://p.scdn.co/mp3-preview/d8140736a6131cb5...      0.9220  115.075   
2  https://p.scdn.co/mp3-preview/c8af28fb15185b18...      0.9440   79.565   
3  https://p.scdn.co/mp3-preview/7629b8e9f31f6e9b...      0.9380  112.822   
4  https://p.scdn.co/mp3-preview/32be593c0eb82868...      0.9430   81.260   

   time_signature                                         track_href  \
0             4.0  https://api.spotify.com/v1/tracks/5qljLQuKnNJf...   
1             3.0  https://api.spotify.com/v1/tracks/3VAX2MJdmdqA...   
2             4.0  https://api.spotify.com/v1/tracks/1L3YAhsEMrGV...   
3             3.0  https://api.spotify.com/v1/tracks/6aCe9zzoZmCo...   
4             4.0  https://api.spotify.com/v1/tracks/1Vo802A38tPF...   

   track_name_prev track_number                                   uri valence  \
0         track_14          1.0  spotify:track:5qljLQuKnNJf4F4vfxQB0V  0.6220   
1          track_3          3.0  spotify:track:3VAX2MJdmdqARLSU5hPMpm  0.5890   
2          track_4          4.0  spotify:track:1L3YAhsEMrGVvCgDXj2TYn  0.0850   
3          track_9          1.0  spotify:track:6aCe9zzoZmCojX7bbgKKtf  0.5330   
4          track_2          2.0  spotify:track:1Vo802A38tPFHmje1h91um  0.0906   

    type  
0  track  
1  track  
2  track  
3  track  
4  track  

[5 rows x 31 columns]

In [46]:
df.columns

Index(['acousticness', 'album_id', 'analysis_url', 'artists_id',
       'available_markets', 'country', 'danceability', 'disc_number',
       'duration_ms', 'energy', 'href', 'id', 'instrumentalness', 'key',
       'liveness', 'loudness', 'lyrics', 'mode', 'name', 'playlist',
       'popularity', 'preview_url', 'speechiness', 'tempo', 'time_signature',
       'track_href', 'track_name_prev', 'track_number', 'uri', 'valence',
       'type'],
      dtype='object')

In [ ]:
album_id = df['album_id']

for row in album_id[10:50]:
    print(row)
    print(type(row))

In [ ]:
artists_id = df['artists_id']

for row in artists_id[10:50]:
    print(row)
    print(type(row))

In [ ]:
lyrics = df['lyrics']

for row in lyrics[10:20]:
    print(row)
    print(type(row))

In [ ]:
id = df['id']

for row in id[10:50]:
    print(row)
    print(type(row))

In [47]:
# drop column: artists_id
df.drop(['artists_id'], axis=1, inplace=True)

In [48]:
df.head()

acousticness                album_id  \
0         0.294  0D3QufeCudpQANOR7luqdr   
1         0.863  1bcqsH5UyTBzmh9YizdsBE   
2         0.750  4tKijjmxGClg4JOLAyo2qE   
3         0.763  6FeJF5r8roonnKraJxr4oB   
4         0.770  4tKijjmxGClg4JOLAyo2qE   

                                        analysis_url  \
0  https://api.spotify.com/v1/audio-analysis/5qlj...   
1  https://api.spotify.com/v1/audio-analysis/3VAX...   
2  https://api.spotify.com/v1/audio-analysis/1L3Y...   
3  https://api.spotify.com/v1/audio-analysis/6aCe...   
4  https://api.spotify.com/v1/audio-analysis/1Vo8...   

                                   available_markets country  danceability  \
0  ['AD', 'AE', 'AR', 'AT', 'AU', 'BE', 'BG', 'BH...      BE         0.698   
1  ['AD', 'AE', 'AR', 'AT', 'AU', 'BE', 'BG', 'BH...      BE         0.719   
2                                             ['GB']      BE         0.466   
3  ['AD', 'AE', 'AR', 'AT', 'AU', 'BE', 'BG', 'BH...      BE         0.719   
4                                             ['GB']      BE         0.460   

   disc_number  duration_ms  energy  \
0          1.0     235584.0   0.606   
1          1.0     656960.0   0.308   
2          1.0     492840.0   0.931   
3          1.0     316578.0   0.126   
4          1.0     558880.0   0.942   

                                                href  ...  \
0  https://api.spotify.com/v1/tracks/5qljLQuKnNJf...  ...   
1  https://api.spotify.com/v1/tracks/3VAX2MJdmdqA...  ...   
2  https://api.spotify.com/v1/tracks/1L3YAhsEMrGV...  ...   
3  https://api.spotify.com/v1/tracks/6aCe9zzoZmCo...  ...   
4  https://api.spotify.com/v1/tracks/1Vo802A38tPF...  ...   

                                         preview_url  speechiness    tempo  \
0  https://p.scdn.co/mp3-preview/1b05a902da3a251d...       0.0262  115.018   
1  https://p.scdn.co/mp3-preview/d8140736a6131cb5...       0.9220  115.075   
2  https://p.scdn.co/mp3-preview/c8af28fb15185b18...       0.9440   79.565   
3  https://p.scdn.co/mp3-preview/7629b8e9f31f6e9b...       0.9380  112.822   
4  https://p.scdn.co/mp3-preview/32be593c0eb82868...       0.9430   81.260   

   time_signature                                         track_href  \
0             4.0  https://api.spotify.com/v1/tracks/5qljLQuKnNJf...   
1             3.0  https://api.spotify.com/v1/tracks/3VAX2MJdmdqA...   
2             4.0  https://api.spotify.com/v1/tracks/1L3YAhsEMrGV...   
3             3.0  https://api.spotify.com/v1/tracks/6aCe9zzoZmCo...   
4             4.0  https://api.spotify.com/v1/tracks/1Vo802A38tPF...   

  track_name_prev  track_number                                   uri valence  \
0        track_14           1.0  spotify:track:5qljLQuKnNJf4F4vfxQB0V  0.6220   
1         track_3           3.0  spotify:track:3VAX2MJdmdqARLSU5hPMpm  0.5890   
2         track_4           4.0  spotify:track:1L3YAhsEMrGVvCgDXj2TYn  0.0850   
3         track_9           1.0  spotify:track:6aCe9zzoZmCojX7bbgKKtf  0.5330   
4         track_2           2.0  spotify:track:1Vo802A38tPFHmje1h91um  0.0906   

    type  
0  track  
1  track  
2  track  
3  track  
4  track  

[5 rows x 30 columns]

In [49]:
df.dtypes

acousticness         float64
album_id              object
analysis_url          object
available_markets     object
country               object
danceability         float64
disc_number          float64
duration_ms          float64
energy               float64
href                  object
id                    object
instrumentalness     float64
key                  float64
liveness             float64
loudness             float64
lyrics                object
mode                 float64
name                  object
playlist              object
popularity           float64
preview_url           object
speechiness          float64
tempo                float64
time_signature       float64
track_href            object
track_name_prev       object
track_number         float64
uri                   object
valence              float64
type                  object
dtype: object

Notable Columns:

In [ ]:
# 'lyrics' column:
df['lyrics'].iloc[0]

"\r\n\r\nPerhaps I am bound to be restless\r\nAlways yearning, never satisfied\r\nPerhaps I'm aching for nothing\r\nThat words can't ever define\r\n\r\nWas I happy in those moments\r\nOr have I made it seem that way?\r\nWas I back there tomorrow\r\nCould feel as lacking as today?\r\n\r\nBut you in your golden hour\r\nAnd you with your sweetest song\r\nAnd you with your tender offerings\r\nHow could I ever do you wrong?\r\n\r\nThere's an unrest that keeps me marching\r\nThere's foolish hope that can be found\r\nSomething so true and lasting\r\nCould it land me on the ground?\r\nStill I like my world boundless\r\nIt never ceases to entice\r\nIt's the hunger in my bones\r\nIt's the new day in my eyes\r\n\r\nAnd you in your darkened hour\r\nAnd me and my broken soul\r\nHere you come with the tender offerings\r\nHow could I ever do you wrong?\r\n\r\nAnd me in my doubting hour\r\nAnd you with your soothing song\r\nAnd you when you call my mercy\r\nAnd you when you can't be strong\r\nHere you

In [ ]:
# Plays the track in a web browser. Want to do something like this in React. Figure out how.
df['preview_url'].iloc[0]	

'https://p.scdn.co/mp3-preview/1b05a902da3a251d07a38aa710ffae559fc33d08?cid=b3cdb16d0df2409abf6a8f6c2f6c2e0c'

In [ ]:
# Stats for the whole table:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101939 entries, 0 to 101938
Data columns (total 31 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   acousticness       101939 non-null  float64
 1   album_id           101939 non-null  object 
 2   analysis_url       101939 non-null  object 
 3   artists_id         101939 non-null  object 
 4   available_markets  101939 non-null  object 
 5   country            101939 non-null  object 
 6   danceability       101939 non-null  float64
 7   disc_number        101939 non-null  float64
 8   duration_ms        101939 non-null  float64
 9   energy             101939 non-null  float64
 10  href               101939 non-null  object 
 11  id                 101939 non-null  object 
 12  instrumentalness   101939 non-null  float64
 13  key                101939 non-null  float64
 14  liveness           101939 non-null  float64
 15  loudness           101939 non-null  float64
 16  ly

From above, we can see tracks.csv has:
101,939 rows, 31 columns.


Individual Column Exploration:

In [ ]:
# stats for 'acousticness' column:
df['acousticness'].describe()

count    101939.000000
mean          0.352124
std           0.334855
min           0.000000
25%           0.040700
50%           0.238000
75%           0.645000
max           0.996000
Name: acousticness, dtype: float64

In [ ]:
df['danceability'].describe()

count    101939.000000
mean          0.586015
std           0.177724
min           0.000000
25%           0.480000
50%           0.610000
75%           0.714000
max           0.989000
Name: danceability, dtype: float64

In [ ]:
df['duration_ms'].describe()

count    1.019390e+05
mean     2.467708e+05
std      1.904303e+05
min      1.155000e+03
25%      1.840000e+05
50%      2.168930e+05
75%      2.610550e+05
max      5.505831e+06
Name: duration_ms, dtype: float64

Export the cleaned file:


In [ ]:
# df.to_csv('clean_tracks.csv') 
# files.download('clean_tracks.csv')